src/appfl/algorithm/client_optimizer.py

```diff
@@ -33,41 +33,64 @@ def update(self):
         self.model.to(self.cfg.device)
 
         optimizer = eval(self.optim)(self.model.parameters(), **self.optim_args)
+ 
 
         """ Multiple local update """
         start_time=time.time()
-        for t in range(self.num_local_epochs):
-
-            if self.cfg.validation == True and self.test_dataloader != None:
-                train_loss, train_accuracy = super(ClientOptim, self).client_validation(
-                    self.dataloader
-                )
-                test_loss, test_accuracy = super(ClientOptim, self).client_validation(
-                    self.test_dataloader
-                )
-                per_iter_time = time.time() - start_time
-                super(ClientOptim, self).client_log_content(
-                    t, per_iter_time, train_loss, train_accuracy, test_loss, test_accuracy
-                )
-                ## return to train mode
-                self.model.train()
+        ## initial evaluation
+        if self.cfg.validation == True and self.test_dataloader != None:
+            test_loss, test_accuracy = super(ClientOptim, self).client_validation(
+                self.test_dataloader
+            )
+            per_iter_time = time.time() - start_time
+            super(ClientOptim, self).client_log_content(
+                0, per_iter_time, 0, 0, test_loss, test_accuracy
+            )
+            ## return to train mode
+            self.model.train()        
 
+        ## local training 
+        for t in range(self.num_local_epochs):
             start_time=time.time()
-            for data, target in self.dataloader:
+            train_loss = 0
+            train_correct = 0            
+            tmptotal = 0
+            for data, target in self.dataloader:                
+                tmptotal += len(target)
                 data = data.to(self.cfg.device)
                 target = target.to(self.cfg.device)
                 optimizer.zero_grad()
                 output = self.model(data)
                 loss = self.loss_fn(output, target)
                 loss.backward()
                 optimizer.step()
+                
+                train_loss += loss.item()
+                if output.shape[1] == 1:
+                    pred = torch.round(output)
+                else:
+                    pred = output.argmax(dim=1, keepdim=True)
+                train_correct += pred.eq(target.view_as(pred)).sum().item()
```

Comment:
@minseok-ryu This part must be fixing the issue #94. Can you put some comments here about the fix?

Reply:
This part is related to the "training loss computation."